### DISCUSSION TITLE
Proposal #1 for a new intern project

### DISCUSSION THREAD

**Discussion by ghost** [BlockchainCommons/Community#40] [Ideas]
![image](https://user-images.githubusercontent.com/13405205/120316185-c35a0000-c2fa-11eb-8ac0-f0bdbf047ffc.png)

## What?

DEX: Decentralized Exchange using Bitcoin technologies

## Why?

1. Less DEXs are available for bitcoiners
2. Its possible to create DEX using Bitcoin tech
3. [Demand for P2P exchanges](https://www.reddit.com/r/Bitcoin/comments/kccyh5/india_overtakes_china_in_peertopeer_bitcoin/) exists in the country I live in (India) and if we create a DEX with good UI/UX, some users may migrate from centralized exchanges considering the trade-offs.
4. Always good to have multiple DEX with different options. Few that already exist include [Bisq](https://bisq.network/). HodlHodl is also famous among Bitcoiners, not used much in India but it cannot be considered decentralized. Its a good P2P exchange though and uses [2-of-3 multisig setup](https://gitlab.com/hodlhodl-public/public_docs/blob/master/multisig_spec.md) in which 1 key belongs to HodlHodl in every trade.
5. No option for Indians to trade Indian stock market using bitcoin

## How?

DEX will have 2 parts: 1. Spot 2. Derivatives

<details><summary>Other options</summary>

<p>There were few more options planned earlier when I started working on this idea with another dev. We had plans to create an android app so providing an option to run full node (pruned) on phone, send bitcoin without internet (using text messages) etc. were also included in it as advanced options. Right now I have decided to work on desktop application using .NET because I don't know Java, Kotlin etc. and other dev isn't interested to work on it anymore.</p></details>

**1. Spot: Buy/Sell bitcoin using fiat**

 - Payment methods: Bank account, Google Pay, Paytm, Gift cards etc.
 - 2-of-3 multisig will be used initially which works similar to HodlHodl
 - HodlHodl uses the payment password to create a new seed. In this project, user will be asked to enter the public key directly which will be used to create a multisig address.
 - Buyer, Seller and one person from our team will hold one key each for every trade initially. Later we can replace our team with anyone who is eligible according to:
 
Holds U number of Unroll tokens, completed T number of trades, trading volume above V, etc.

The list of all eligible users will be available in the application, buyer and seller can agree on one of these to hold the 3rd key for trade. This key will be helpful if there are any disputes during the trade.

**2. Derivatives: Long/Short few markets using bitcoin**

 - Different markets include: 1. BTCUSD 2. Indian stock market 3. Gold 4. [Fees per block](https://www.youtube.com/watch?v=rBCG0btUlTw) 5. Hashrate 6. Prediction markets (Soft forks, Sports, Important events etc.) 7. [Copy trading](https://gist.github.com/prayank23/4995d251c3ff85dd6ee3f8031c1be2ca) (using multi oracles)
 - Oracles required for different markets will be implemented using [discreet log contracts](https://github.com/discreetlogcontracts/dlcspecs)
 - L-BTC and a token for Unroll issued on [Liquid](https://docs.blockstream.com/liquid/technical_overview.html) will be used. L-BTC for collateral and Unroll token for fees, incentives etc.

_Example of P2P derivatives project using DLC:_ https://github.com/p2pderivatives/p2pderivatives-client

### What have I done until now?

1. Research
2. Discuss the idea with different people
3. Experiment with [Bitcoin Core RPCs](https://bitcoincore.org/en/doc/0.21.0/rpc/). Learn Bitcoin from Command Line by Blockchain Commons is very helpful for beginners: https://github.com/BlockchainCommons/Learning-Bitcoin-from-the-Command-Line
4. Experiment with Caravan (Stateless Multisig Coordinator): https://github.com/unchained-capital/caravan
5. Learn more about discreet log contracts and follow development
6. Issue assets on Liquid 
7. Started creating application using C#. Working on login and spot market first.
8. I want to implement authentication using LN-URL so that anyone with supported LN wallet can login. There is one LN-URL library for .NET but I have not implemented it yet: https://github.com/fiatjaf/awesome-lnurl#libraries
9. As its a DEX project we cannot use centralized server for anything. I am looking for some distributed database API for C#. Initially wanted to use torrents but the [libraries I found](https://twitter.com/prayankgahlot/status/1389449217110859777?s=20) had no proper documentation. "nostr" by fiatjaf looks good but there is no .NET library: https://github.com/fiatjaf/nostr#libraries. Condensation DB is also an interesting project but only implemented in JavaScript and Java right now: https://github.com/CondensationDB

---

**Since its a big project I have divided it in to few parts:**

1. Login (LNURL-AUTH)
2. Distributed Database
3. Spot market using 2-of-3 multsig
4. Derivatives (Discreet log contracts)

Implementing P2P derivatives for Indian Stock market or/and market for hedging fees should be achievable during the internship using [NDLC](https://github.com/dgarage/NDLC) library in a desktop application with C# Cc: @jesseposner

**Comment by nochiel** [comment]
> 1. Spot: Buy/Sell bitcoin using fiat

I think you should omit this part. Handling fiat is, in general, a regulatory landmine/honeypot, as far as I am aware. The legal and biz dev overhead it would impose on a project is simply not worth it in a world where stablecoins exist. Instead, it might be easier to require a stablecoin to trade.

---

I was thinking about what your minimum viable product might look like and I found [this thread](https://twitter.com/hasufl/status/1394268563012268033): 

> If you want to kickstart Defi on Bitcoin/LN, **focus on a single usecase**:  Leverage for Traders.
> This requires two building blocks:
> 1) a lending protocol where people can lock BTC and borrow a stablecoin against it (can be synthetic or centralized)
> 2) a DEX to exchange them

(emphasis mine)

Maybe:
- The trading protocol can be a very simplified version of [dydx's solo](https://github.com/dydxprotocol/solo).
- The lending protocol could be a simplified version of [aave's](https://github.com/aave/aave-protocol).

---

A DEX on Lightning ⚡ is necessarily an L3? (cf. ["The Bitcoin Stack"](https://www.youtube.com/watch?v=qlV5_udJkC0)). But:
- Bitcoin DEXs like <http://sovryn.app> seem to have chosen to build on the [RSK sidechain](https://www.rsk.co/rsk-blockchain/). 
- Do you want a DEX on LN without relying on sidechains like RSK or [Stacks](https://www.stacks.co/what-is-stacks)? Is it easier to prototype on one of them?

**Reply by ghost** [reply]
@nochiel Thanks for your feedback about the proposal :)

> I think you should omit this part. Handling fiat is, in general, a regulatory landmine/honeypot, as far as I am aware. The legal and biz dev overhead it would impose on a project is simply not worth it in a world where stablecoins exist. Instead, it might be easier to require a stablecoin to trade.

1. **Non-KYC** peer to peer exchanges are important for privacy
2. Using stablecoins is not a solution because users will need to buy stablecoins from an exchange. Onboarding from fiat to bitcoin should not require KYC to improve privacy in the ecosystem.
3. Regulatory issues can be avoided if project is decentralized and peers can be selected manually similar to Bisq based on their trades, feedback, security deposit etc.

> The trading protocol can be a very simplified version of dydx's solo.
The lending protocol could be a simplified version of aave's.

dYdX is a derivatives exchange so the derivatives part in this project can be similar. Lending is not planned yet but will be a good feature if added in future versions. HodlHodl has implemnted P2P lending which can be helpful in understanding the things involved.

> Bitcoin DEXs like http://sovryn.app seem to have chosen to build on the RSK sidechain.

Yes its a good project and I have tried it. Although I don't want to use solidity or RSK in this project.

> Do you want a DEX on LN without relying on sidechains like RSK or Stacks? Is it easier to prototype on one of them?

Sorry I have not researched enough about LN and mostly experiment with on-chain projects or Sidechain. Liquid is one of the sidechains that I tried and found easy to use, develop projects etc. Its an implementation of [Elements](https://github.com/ElementsProject/elements)

There are lot of projects which are using LN, however I think [Omni BOLT](https://github.com/omnilaboratory/OmniBOLT-spec) would work better if it works as expected.

**Reply by nochiel** [reply]
> Non-KYC peer to peer exchanges are important for privacy

Absolutely. A pure p2p exchange like Bisq(or HodlHodl , LocalCryptos?) is ideal. I misunderstood what you had in mind and was worried about a situation I've seen, where government prohibits payment processors from processing payments to a registered/unregistered fiat onramps. 

Thanks for the clarification!
